File: services/elitea-main/internal/llmproxy/identity.go
// Package llmproxy is elitea-main's lightweight /llm gateway role: a streaming
// reverse proxy that authenticates at the edge, resolves the caller's project,
// injects signed identity headers, and byte-streams /llm to
// elitea-llm-gateway-svc over mTLS (design §2, ADR-0015 Option E). It does NOT
// import bifrost/core, translate wire dialects, or run governance.
package llmproxy

import (
	"context"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"net/http"
	"strconv"

	"github.com/EliteaAI/elitea-platform/services/elitea-main/internal/api/middleware"
	"github.com/EliteaAI/elitea-platform/services/elitea-main/internal/auth"
	"github.com/EliteaAI/elitea-platform/services/elitea-main/internal/infra/db/tenant"
)

// Signed internal identity headers forwarded to the gateway. The gateway trusts
// these only because the hop is mTLS-internal (design §2, §6.1); the HMAC
// signature is defense-in-depth so a header that leaks onto the internal network
// without the shared secret cannot forge a project identity.
const (
	HeaderProjectID = "X-Elitea-Project-Id"
	HeaderUserID    = "X-Elitea-User-Id"
	HeaderTenantID  = "X-Elitea-Tenant-Id"
	// HeaderExecutionID names the runtime execution a /llm request was made
	// FROM — the value elitea_runtime.execution_jobs is keyed by. It is what
	// gives the gateway's request log and usage ledger an AGENT dimension.
	//
	// It is an EXECUTION id and not an agent id on purpose: the runtime already
	// mints it, it is stable across retries of the same turn, and resolving it
	// to an agent at READ time keeps execution_jobs' two project columns —
	// resource_project_id and projection_project_id, which can differ — out of
	// the log. Copying an agent id onto the log would copy that ambiguity with
	// it (internal/infra/db/repos/analytics.go).
	//
	// Unlike the three headers above it is not resolved from this process's own
	// middleware: the caller that knows it is the runtime worker, which sends
	// it inbound. The edge therefore READS it off the inbound request, validates
	// its shape, and re-emits it as part of the signed tuple — see
	// executionIDFromHeader.
	HeaderExecutionID = "X-Elitea-Execution-Id"
	// HeaderSignature carries "sha256=<hex>" over the canonical identity tuple.
	HeaderSignature = "X-Elitea-Identity-Signature"
)

// The canonical string is prefixed with a scheme VERSION so it can be rotated
// without ambiguity, and this is the rotation the prefix was put there for.
//
// v1 signs (project, user, tenant). v2 signs (project, user, tenant, execution).
//
// THE CANONICAL STRING IS DUPLICATED ACROSS TWO INDEPENDENTLY DEPLOYED GO
// MODULES — this file signs it and services/elitea-llm-gateway's
// internal/llmproxy/identity.go verifies it. Changing it in place would fail
// EVERY gateway request in both directions for the length of a rolling deploy.
//
// So this edge signs v2 only when it actually has an execution id, and v1 —
// byte for byte what it signed before — otherwise. Every existing caller (chat,
// the SDK, anything that is not a runtime execution) therefore keeps producing
// a signature an OLD gateway still accepts, and only the new agent-tagged
// requests need a new gateway. The deploy order is "gateway first" rather than
// a hard cutover, and no request is signed under a scheme its peer cannot at
// least recognise.
const (
	signatureVersionV1 = "v1"
	signatureVersionV2 = "v2"
)

// maxExecutionIDLen bounds the inbound execution id.
//
// The value is minted by the runtime and is a short opaque token; the bound
// exists because this one field arrives on the request rather than being
// resolved from the session, so it is the only part of the signed tuple a
// caller can put bytes into. See executionIDFromHeader for what it may contain
// and why an invalid value is DROPPED rather than refused.
const maxExecutionIDLen = 128

// identity is the resolved edge identity forwarded to the gateway. The
// projectID is the resolved, non-secret handle used as the Bifrost virtual-key
// value at the gateway — never the raw Elitea key (design §2, §6.1).
type identity struct {
	projectID string
	userID    string
	tenantID  string
	// executionID is the runtime execution this request belongs to, empty for
	// every request that is not made from one.
	executionID string
}

// identityFromContext derives the forwarded identity from the request context
// populated by the auth, project-resolution, and tenant middleware. Any field
// that cannot be resolved is left empty; the proxy still forwards, and the
// gateway decides whether a missing project is fatal (IsVkMandatory).
func identityFromContext(ctx context.Context) identity {
	var id identity

	if pc, ok := middleware.ProjectFromContext(ctx); ok && pc.ProjectID > 0 {
		id.projectID = strconv.Itoa(pc.ProjectID)
	}
	if u, ok := auth.UserFromContext(ctx); ok {
		id.userID = u.ID
	}
	if t, ok := tenant.TenantFromContext(ctx); ok {
		id.tenantID = t
	}
	return id
}

// canonical returns the deterministic string signed by the HMAC. Newline
// separation with a version prefix prevents field-concatenation ambiguity
// (e.g. project "1"+user "2" colliding with project "12").
func (id identity) canonical(version string) string {
	base := version + "\n" + id.projectID + "\n" + id.userID + "\n" + id.tenantID
	if version == signatureVersionV2 {
		return base + "\n" + id.executionID
	}
	return base
}

// signatureVersion is the scheme this identity signs under: v2 exactly when
// there is an execution id to cover, v1 otherwise.
func (id identity) signatureVersion() string {
	if id.executionID != "" {
		return signatureVersionV2
	}
	return signatureVersionV1
}

// sign returns the "sha256=<hex>" signature of the identity under secret.
func (id identity) sign(secret []byte) string {
	return id.signVersion(secret, id.signatureVersion())
}

func (id identity) signVersion(secret []byte, version string) string {
	mac := hmac.New(sha256.New, secret)
	mac.Write([]byte(id.canonical(version)))
	return "sha256=" + hex.EncodeToString(mac.Sum(nil))
}

// executionIDFromHeader reads and validates the inbound execution id.
//
// An INVALID value yields "", which drops the dimension for that request rather
// than refusing it. That is deliberate: the execution id decorates analytics
// and decides nothing — not routing, not authorization, not billing — so a
// malformed one costs a row in a breakdown, while refusing would let a bad
// header break the product's most visible path.
//
// The charset is restrictive because the value is re-emitted as a HEADER and
// then signed: a CR or LF in it would be header injection on the outbound
// request, and nothing outside this set belongs in an id the runtime minted.
//
// A caller CAN put an id here for an execution it did not make. That is
// contained by resolving the id at READ time inside the project the log row
// already names (analytics.go): an id from another project resolves to nothing,
// and one from the caller's own project attributes to an agent that caller can
// already see. It is never an authorization input.
func executionIDFromHeader(h http.Header) string {
	value := h.Get(HeaderExecutionID)
	if value == "" || len(value) > maxExecutionIDLen {
		return ""
	}
	for i := 0; i < len(value); i++ {
		c := value[i]
		switch {
		case c >= 'a' && c <= 'z', c >= 'A' && c <= 'Z', c >= '0' && c <= '9':
		case c == '-', c == '_', c == '.', c == ':':
		default:
			return ""
		}
	}
	return value
}

// stripIdentityHeaders removes any caller-supplied authentication and identity
// headers before the request is forwarded to the gateway. The gateway must only
// ever see the signed X-Elitea-* identity that this edge process injects; it
// must never see client-supplied authentication material.
//
// Stripped headers:
//   - X-Elitea-* (signed identity injected below; strip first to avoid leaking
//     any client-spoofed value)
//   - X-Auth-Type / X-Auth-Id / X-Auth-Reference (Traefik forward-auth headers)
//   - Authorization, X-Api-Key (bearer / API-key credentials)
//   - Cookie (session cookies; must not reach the downstream gateway)
//   - X-Project-Id / OpenAI-Organization (the edge project selector; the edge
//     consumes it and speaks the result as signed identity)
//   - OpenAI-Project (never a selector, but it must not travel onward either)
func stripIdentityHeaders(h http.Header) {
	// Signed edge identity headers — re-injected after stripping.
	h.Del(HeaderProjectID)
	h.Del(HeaderUserID)
	h.Del(HeaderTenantID)
	h.Del(HeaderExecutionID)
	h.Del(HeaderSignature)

	// The project headers are edge control headers (issue #318). The edge
	// already read the selector, checked the caller's membership, and expressed
	// the result as X-Elitea-Project-Id. Forwarding any of them would send an
	// Elitea project id onward under a name a real provider reads: OpenAI
	// rejects a request whose OpenAI-Organization names an organization the key
	// cannot use. OpenAI-Project is stripped as well, although the edge never
	// reads it as a selector, because it carries an Elitea project id too.
	for _, name := range middleware.ProjectHeadersStrippedOutbound() {
		h.Del(name)
	}

	// Traefik forward-auth headers that the auth middleware reads; remove so the
	// gateway never sees inbound authentication context.
	h.Del("X-Auth-Type")
	h.Del("X-Auth-Id")
	h.Del("X-Auth-Reference")

	// Standard HTTP authentication material.
	h.Del("Authorization")
	h.Del("X-Api-Key")
	h.Del("Cookie")
}

// injectIdentity strips any client-supplied identity headers on out, then sets
// the edge-resolved identity. When secret is non-empty and a project is present
// it also attaches the HMAC signature.
func injectIdentity(ctx context.Context, out http.Header, secret []byte) {
	// READ BEFORE STRIP. stripIdentityHeaders deletes every X-Elitea-* header,
	// this one included, and `out` is a copy of the inbound headers — so the
	// caller-supplied execution id has to be taken (and validated) here, before
	// the strip removes it. Taking it afterwards would silently produce an empty
	// dimension on every request: a column that is always NULL and a breakdown
	// that is always empty, with nothing failing.
	executionID := executionIDFromHeader(out)

	stripIdentityHeaders(out)

	id := identityFromContext(ctx)
	id.executionID = executionID
	if id.projectID != "" {
		out.Set(HeaderProjectID, id.projectID)
	}
	if id.userID != "" {
		out.Set(HeaderUserID, id.userID)
	}
	if id.tenantID != "" {
		out.Set(HeaderTenantID, id.tenantID)
	}
	if id.executionID != "" {
		out.Set(HeaderExecutionID, id.executionID)
	}

	// Sign only when we have both a secret and a project identity to bind; an
	// unsigned request over the mTLS-internal network is still authenticated by
	// the transport, but a signature lets the gateway reject stray traffic.
	if len(secret) > 0 && id.projectID != "" {
		out.Set(HeaderSignature, id.sign(secret))
	}
}

// SignIdentityHeaders sets the signed X-Elitea-* identity headers on out from
// explicit project/user/tenant values, using the identical HMAC scheme
// injectIdentity uses. It exists for elitea-main callers that resolve their
// own project identity from something other than the request-scoped
// middleware context injectIdentity reads — e.g. the configurations
// check-connection client (internal/api/v2/configurations), whose project id
// is a URL path parameter, not a *middleware.ProjectContext the /configurations
// mount never populates. Mirrors the gateway's own exported
// SignIdentityHeaders (elitea-llm-gateway/internal/llmproxy/identity.go) so
// the two sides cannot silently diverge on the signing scheme.
func SignIdentityHeaders(out http.Header, secret []byte, projectID, userID, tenantID, executionID string) {
	stripIdentityHeaders(out)
	id := identity{projectID: projectID, userID: userID, tenantID: tenantID, executionID: executionID}
	if id.projectID != "" {
		out.Set(HeaderProjectID, id.projectID)
	}
	if id.userID != "" {
		out.Set(HeaderUserID, id.userID)
	}
	if id.tenantID != "" {
		out.Set(HeaderTenantID, id.tenantID)
	}
	if id.executionID != "" {
		out.Set(HeaderExecutionID, id.executionID)
	}
	if len(secret) > 0 && id.projectID != "" {
		out.Set(HeaderSignature, id.sign(secret))
	}
}

// verifyIdentitySignature reports whether the identity headers on h carry a
// valid signature under secret. It is used by tests here and is safe for the
// gateway side to mirror. Returns false if the signature header is absent.
func verifyIdentitySignature(h http.Header, secret []byte) bool {
	got := h.Get(HeaderSignature)
	if got == "" {
		return false
	}
	id := identity{
		projectID:   h.Get(HeaderProjectID),
		userID:      h.Get(HeaderUserID),
		tenantID:    h.Get(HeaderTenantID),
		executionID: h.Get(HeaderExecutionID),
	}
	// Mirrors the gateway's acceptance rule exactly: v2 always, v1 only when no
	// execution id is present. A v1 canonical string does not cover the id, so
	// accepting one beside the header would make the id caller-attachable —
	// the single thing signing it prevents.
	if hmac.Equal([]byte(got), []byte(id.signVersion(secret, signatureVersionV2))) {
		return true
	}
	if id.executionID != "" {
		return false
	}
	return hmac.Equal([]byte(got), []byte(id.signVersion(secret, signatureVersionV1)))
}

File: services/elitea-deepwiki/src/elitea_deepwiki/security/identity.py
"""Signed identity headers — the third implementation of one scheme.

ADR-0022 decision 5 has elitea-main proxy to this service "over mTLS with
HMAC-signed identity headers, the established llmproxy pattern". That pattern
is already written twice in Go:

    services/elitea-main/internal/llmproxy/identity.go        signs
    services/elitea-llm-gateway/internal/llmproxy/identity.go verifies

This is the third, and it must agree with them byte for byte or every request
from the facade fails. The canonical string, the header names and the
version-fallback rule below are transcribed from those files deliberately and
are not free to "improve":

    canonical(v1) = "v1\\n" + project + "\\n" + user + "\\n" + tenant
    canonical(v2) = canonical-body-of-v1-with-"v2" + "\\n" + execution
    signature     = "sha256=" + hex(HMAC-SHA256(secret, canonical))

The Go comment explains why the version prefix exists: the string is duplicated
across independently deployed modules, so changing it in place would fail every
request in both directions for the length of a rolling deploy. The same now
applies here, with a third module in the rotation.

**Verification accepts v2 always, and v1 only when no execution id is
present.** Falling back to v1 for a request that carries an execution id would
make that id caller-attachable, which is the thing signing it prevents.
"""

from __future__ import annotations

import hmac
import logging
from dataclasses import dataclass
from hashlib import sha256

logger = logging.getLogger(__name__)

HEADER_PROJECT_ID = "x-elitea-project-id"
HEADER_USER_ID = "x-elitea-user-id"
HEADER_TENANT_ID = "x-elitea-tenant-id"
HEADER_EXECUTION_ID = "x-elitea-execution-id"
HEADER_SIGNATURE = "x-elitea-identity-signature"

#: Every identity header a client must not be able to set for itself. The edge
#: resolves these; a value arriving from outside is a spoofing attempt, and the
#: middleware strips them before anything downstream can read one.
IDENTITY_HEADERS = (
    HEADER_PROJECT_ID,
    HEADER_USER_ID,
    HEADER_TENANT_ID,
    HEADER_EXECUTION_ID,
    HEADER_SIGNATURE,
)

SIGNATURE_VERSION_V1 = "v1"
SIGNATURE_VERSION_V2 = "v2"


@dataclass(frozen=True)
class Identity:
    """The resolved caller, as the facade signed it."""

    project_id: str = ""
    user_id: str = ""
    tenant_id: str = ""
    execution_id: str = ""

    def canonical(self, version: str) -> str:
        base = f"{version}\n{self.project_id}\n{self.user_id}\n{self.tenant_id}"
        if version == SIGNATURE_VERSION_V2:
            return f"{base}\n{self.execution_id}"
        return base

    def signature_version(self) -> str:
        return SIGNATURE_VERSION_V2 if self.execution_id else SIGNATURE_VERSION_V1

    def sign(self, secret: bytes, version: str | None = None) -> str:
        version = version or self.signature_version()
        digest = hmac.new(secret, self.canonical(version).encode(), sha256)
        return "sha256=" + digest.hexdigest()

    def is_empty(self) -> bool:
        return not (self.project_id or self.user_id or self.tenant_id)


def identity_from_headers(headers) -> Identity:
    """Read an identity off a Starlette/ASGI header mapping (case-insensitive)."""
    return Identity(
        project_id=headers.get(HEADER_PROJECT_ID, "") or "",
        user_id=headers.get(HEADER_USER_ID, "") or "",
        tenant_id=headers.get(HEADER_TENANT_ID, "") or "",
        execution_id=headers.get(HEADER_EXECUTION_ID, "") or "",
    )


def verify_signature(headers, secret: bytes) -> bool:
    """Whether the identity headers carry a valid signature under ``secret``.

    An empty secret disables verification, matching the Go verifier: the mTLS
    transport still authenticates the hop, and the edge only signs when a
    secret is configured. A configured secret with a missing or wrong
    signature is a refusal.
    """
    if not secret:
        return True

    got = headers.get(HEADER_SIGNATURE, "")
    if not got:
        return False

    identity = identity_from_headers(headers)

    if hmac.compare_digest(got, identity.sign(secret, SIGNATURE_VERSION_V2)):
        return True

    if identity.execution_id:
        # A v1 signature does not cover the execution id, so accepting one here
        # would make that id caller-attachable — precisely what signing it
        # prevents. Transcribed from the Go verifier, including this refusal.
        return False

    return hmac.compare_digest(got, identity.sign(secret, SIGNATURE_VERSION_V1))


def sign_headers(identity: Identity, secret: bytes) -> dict[str, str]:
    """Produce the headers a facade would send. Used by tests and the harness."""
    headers = {}
    if identity.project_id:
        headers[HEADER_PROJECT_ID] = identity.project_id
    if identity.user_id:
        headers[HEADER_USER_ID] = identity.user_id
    if identity.tenant_id:
        headers[HEADER_TENANT_ID] = identity.tenant_id
    if identity.execution_id:
        headers[HEADER_EXECUTION_ID] = identity.execution_id
    if secret:
        headers[HEADER_SIGNATURE] = identity.sign(secret)
    return headers
